To run this, press "*Runtime*" and press "*Run all*" on a **free** Tesla T4 Google Colab instance!
<div class="align-center">
<a href="https://unsloth.ai/"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
<a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord button.png" width="145"></a>
<a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a></a> Join Discord if you need help + ⭐ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐
</div>

To install Unsloth your local device, follow [our guide](https://docs.unsloth.ai/get-started/install-and-update). This notebook is licensed [LGPL-3.0](https://github.com/unslothai/notebooks?tab=LGPL-3.0-1-ov-file#readme).

You will learn how to do [data prep](#Data), how to [train](#Train), how to [run the model](#Inference), & [how to save it](#Save)


### News


Introducing FP8 precision training for faster RL inference. [Read Blog](https://docs.unsloth.ai/new/fp8-reinforcement-learning).

Unsloth's [Docker image](https://hub.docker.com/r/unsloth/unsloth) is here! Start training with no setup & environment issues. [Read our Guide](https://docs.unsloth.ai/new/how-to-train-llms-with-unsloth-and-docker).

[gpt-oss RL](https://docs.unsloth.ai/new/gpt-oss-reinforcement-learning) is now supported with the fastest inference & lowest VRAM. Try our [new notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/gpt-oss-(20B)-GRPO.ipynb) which creates kernels!

Introducing [Vision](https://docs.unsloth.ai/new/vision-reinforcement-learning-vlm-rl) and [Standby](https://docs.unsloth.ai/basics/memory-efficient-rl) for RL! Train Qwen, Gemma etc. VLMs with GSPO - even faster with less VRAM.

Visit our docs for all our [model uploads](https://docs.unsloth.ai/get-started/all-our-models) and [notebooks](https://docs.unsloth.ai/get-started/unsloth-notebooks).


### Installation

In [ ]:
%%capture
import os, importlib.util
!pip install --upgrade -qqq uv
if importlib.util.find_spec("torch") is None or "COLAB_" in "".join(os.environ.keys()):
    try: import numpy, PIL; get_numpy = f"numpy=={numpy.__version__}"; get_pil = f"pillow=={PIL.__version__}"
    except: get_numpy = "numpy"; get_pil = "pillow"
    !uv pip install -qqq \
        "torch>=2.8.0" "triton>=3.4.0" {get_numpy} {get_pil} torchvision bitsandbytes "transformers==4.56.2" \
        "unsloth_zoo[base] @ git+https://github.com/unslothai/unsloth-zoo" \
        "unsloth[base] @ git+https://github.com/unslothai/unsloth" \
        git+https://github.com/triton-lang/triton.git@0add68262ab0a2e33b84524346cb27cbb2787356#subdirectory=python/triton_kernels
elif importlib.util.find_spec("unsloth") is None:
    !uv pip install -qqq unsloth
!uv pip install --upgrade --no-deps transformers==4.56.2 tokenizers trl==0.22.2 unsloth unsloth_zoo

### Unsloth

We're about to demonstrate the power of the new OpenAI GPT-OSS 20B model through a finetuning example. To use our `MXFP4` inference example, use this [notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/GPT_OSS_MXFP4_(20B)-Inference.ipynb) instead.

In [ ]:
from unsloth import FastLanguageModel
import torch
max_seq_length = 1024
dtype = None

# 4bit pre quantized models we support for 4x faster downloading + no OOMs.
fourbit_models = [
    "unsloth/gpt-oss-20b-unsloth-bnb-4bit", # 20B model using bitsandbytes 4bit quantization
    "unsloth/gpt-oss-120b-unsloth-bnb-4bit",
    "unsloth/gpt-oss-20b", # 20B model using MXFP4 format
    "unsloth/gpt-oss-120b",
] # More models at https://huggingface.co/unsloth

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "unsloth/gpt-oss-20b",
    dtype = dtype, # None for auto detection
    max_seq_length = max_seq_length, # Choose any for long context!
    load_in_4bit = True,  # 4 bit quantization to reduce memory
    full_finetuning = False, # [NEW!] We have full finetuning now!
    # token = "hf_...", # use one if using gated models
)

🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.
🦥 Unsloth Zoo will now patch everything to make training faster!
==((====))==  Unsloth 2025.8.5: Fast Gpt_Oss patching. Transformers: 4.56.0.dev0.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.8.0+cu128. CUDA: 7.5. CUDA Toolkit: 12.8. Triton: 3.4.0
\        /    Bfloat16 = FALSE. FA [Xformers = None. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!
Unsloth: Using float16 precision for gpt_oss won't work! Using float32.


model.safetensors.index.json: 0.00B [00:00, ?B/s]

model-00001-of-00004.safetensors:   0%|          | 0.00/4.00G [00:00<?, ?B/s]

model-00002-of-00004.safetensors:   0%|          | 0.00/4.00G [00:00<?, ?B/s]

model-00003-of-00004.safetensors:   0%|          | 0.00/3.37G [00:00<?, ?B/s]

model-00004-of-00004.safetensors:   0%|          | 0.00/1.16G [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/4 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/165 [00:00<?, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

tokenizer.json:   0%|          | 0.00/27.9M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/446 [00:00<?, ?B/s]

chat_template.jinja: 0.00B [00:00, ?B/s]

We now add LoRA adapters for parameter efficient finetuning - this allows us to only efficiently train 1% of all parameters.

In [ ]:
model = FastLanguageModel.get_peft_model(
    model,
    r = 8, # Choose any number > 0 ! Suggested 8, 16, 32, 64, 128
    target_modules = ["q_proj", "k_proj", "v_proj", "o_proj",
                      "gate_proj", "up_proj", "down_proj",],
    lora_alpha = 16,
    lora_dropout = 0, # Supports any, but = 0 is optimized
    bias = "none",    # Supports any, but = "none" is optimized
    # [NEW] "unsloth" uses 30% less VRAM, fits 2x larger batch sizes!
    use_gradient_checkpointing = "unsloth", # True or "unsloth" for very long context
    random_state = 3407,
    use_rslora = False,  # We support rank stabilized LoRA
    loftq_config = None, # And LoftQ
)

Unsloth: Making `model.base_model.model.model` require gradients


### Reasoning Effort
The `gpt-oss` models from OpenAI include a feature that allows users to adjust the model's "reasoning effort." This gives you control over the trade-off between the model's performance and its response speed (latency) which by the amount of token the model will use to think.

----

The `gpt-oss` models offer three distinct levels of reasoning effort you can choose from:

* **Low**: Optimized for tasks that need very fast responses and don't require complex, multi-step reasoning.
* **Medium**: A balance between performance and speed.
* **High**: Provides the strongest reasoning performance for tasks that require it, though this results in higher latency.

In [ ]:
from transformers import TextStreamer

messages = [
    {"role": "user", "content": "Solve x^5 + 3x^4 - 10 = 3."},
]
inputs = tokenizer.apply_chat_template(
    messages,
    add_generation_prompt = True,
    return_tensors = "pt",
    return_dict = True,
    reasoning_effort = "low", # **NEW!** Set reasoning effort to low, medium or high
).to("cuda")

_ = model.generate(**inputs, max_new_tokens = 64, streamer = TextStreamer(tokenizer))

<|start|>system<|message|>You are ChatGPT, a large language model trained by OpenAI.
Knowledge cutoff: 2024-06
Current date: 2025-08-13

Reasoning: low

# Valid channels: analysis, commentary, final. Channel must be included for every message.
Calls to these tools must go to the commentary channel: 'functions'.<|end|><|start|>user<|message|>Solve x^5 + 3x^4 - 10 = 3.<|end|><|start|>assistant<|channel|>analysis<|message|>Equation: x^5 + 3x^4 - 10 = 3. So x^5 + 3x^4 - 13 =0. Solve for real roots? maybe numeric. Let's try approximate.

We can test integer roots: try x=1 => 1+3


Changing the `reasoning_effort` to `medium` will make the model think longer. We have to increase the `max_new_tokens` to occupy the amount of the generated tokens but it will give better and more correct answer

In [ ]:
from transformers import TextStreamer

messages = [
    {"role": "user", "content": "Solve x^5 + 3x^4 - 10 = 3."},
]
inputs = tokenizer.apply_chat_template(
    messages,
    add_generation_prompt = True,
    return_tensors = "pt",
    return_dict = True,
    reasoning_effort = "medium", # **NEW!** Set reasoning effort to low, medium or high
).to("cuda")

_ = model.generate(**inputs, max_new_tokens = 64, streamer = TextStreamer(tokenizer))

<|start|>system<|message|>You are ChatGPT, a large language model trained by OpenAI.
Knowledge cutoff: 2024-06
Current date: 2025-08-13

Reasoning: medium

# Valid channels: analysis, commentary, final. Channel must be included for every message.
Calls to these tools must go to the commentary channel: 'functions'.<|end|><|start|>user<|message|>Solve x^5 + 3x^4 - 10 = 3.<|end|><|start|>assistant<|channel|>analysis<|message|>The user: "Solve x^5 + 3x^4 - 10 = 3." Wait maybe it's an equation: x^5 + 3x^4 - 10 = 3. The variable x unknown. Solve for x. We need to solve the equation:

x^


Lastly we will test it using `reasoning_effort` to `high`

In [ ]:
from transformers import TextStreamer

messages = [
    {"role": "user", "content": "Solve x^5 + 3x^4 - 10 = 3."},
]
inputs = tokenizer.apply_chat_template(
    messages,
    add_generation_prompt = True,
    return_tensors = "pt",
    return_dict = True,
    reasoning_effort = "high", # **NEW!** Set reasoning effort to low, medium or high
).to("cuda")

_ = model.generate(**inputs, max_new_tokens = 64, streamer = TextStreamer(tokenizer))

<|start|>system<|message|>You are ChatGPT, a large language model trained by OpenAI.
Knowledge cutoff: 2024-06
Current date: 2025-08-13

Reasoning: high

# Valid channels: analysis, commentary, final. Channel must be included for every message.
Calls to these tools must go to the commentary channel: 'functions'.<|end|><|start|>user<|message|>Solve x^5 + 3x^4 - 10 = 3.<|end|><|start|>assistant<|channel|>analysis<|message|>We need to solve the equation: x^5 + 3x^4 - 10 = 3. Or maybe it's x^5 + 3x^4 - 10 = 3? That seems like a polynomial equation: x^5 + 3x^4 - 10


<a name="Data"></a>
### Data Prep

The `HuggingFaceH4/Multilingual-Thinking` dataset will be utilized as our example. This dataset, available on Hugging Face, contains reasoning chain-of-thought examples derived from user questions that have been translated from English into four other languages. It is also the same dataset referenced in OpenAI's [cookbook](https://cookbook.openai.com/articles/gpt-oss/fine-tune-transfomers) for fine-tuning. The purpose of using this dataset is to enable the model to learn and develop reasoning capabilities in these four distinct languages.

In [ ]:
def formatting_prompts_func(examples):
    convos = examples["messages"]
    texts = [tokenizer.apply_chat_template(convo, tokenize = False, add_generation_prompt = False) for convo in convos]
    return { "text" : texts, }

from datasets import load_dataset

def formatting_prompts_func(examples):
    convos = examples["messages"]
    texts = [
        tokenizer.apply_chat_template(
            convo,
            tokenize=False,
            add_generation_prompt=False
        )
        for convo in convos
    ]
    return {"text": texts}

dataset = load_dataset(
    "json",
    data_files={"train": "nz_sovereign_train_v1.jsonl"},
    split="train",
)

dataset = dataset.map(formatting_prompts_func, batched=True, remove_columns=dataset.column_names)
dataset

README.md: 0.00B [00:00, ?B/s]

data/train-00000-of-00001.parquet:   0%|          | 0.00/5.29M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/1000 [00:00<?, ? examples/s]

Dataset({
    features: ['reasoning_language', 'developer', 'user', 'analysis', 'final', 'messages'],
    num_rows: 1000
})

To format our dataset, we will apply our version of the GPT OSS prompt

Let's take a look at the dataset, and check what the 1st example shows

What is unique about GPT-OSS is that it uses OpenAI [Harmony](https://github.com/openai/harmony) format which support conversation structures, reasoning output, and tool calling.

<a name="Train"></a>
### Train the model
Now let's train our model. We do 60 steps to speed things up, but you can set `num_train_epochs=1` for a full run, and turn off `max_steps=None`.

In [ ]:
from trl import SFTConfig, SFTTrainer
trainer = SFTTrainer(
    model = model,
    tokenizer = tokenizer,
    train_dataset = dataset,
    args = SFTConfig(
        per_device_train_batch_size = 1,
        gradient_accumulation_steps = 4,
        warmup_steps = 5,
        num_train_epochs = 1, # Set this for 1 full training run.
        max_steps = 30,
        learning_rate = 2e-4,
        logging_steps = 1,
        optim = "adamw_8bit",
        weight_decay = 0.001,
        lr_scheduler_type = "linear",
        seed = 3407,
        output_dir = "outputs",
        report_to = "none", # Use TrackIO/WandB etc
    ),
)

Unsloth: Switching to float32 training since model cannot work with float16


Unsloth: Tokenizing ["text"] (num_proc=2):   0%|          | 0/1000 [00:00<?, ? examples/s]

We also use Unsloth's `train_on_completions` method to only train on the assistant outputs and ignore the loss on the user's inputs. This helps increase accuracy of finetunes and lower loss as well!

In [ ]:
from unsloth.chat_templates import train_on_responses_only

gpt_oss_kwargs = dict(instruction_part = "<|start|>user<|message|>", response_part="<|start|>assistant<|channel|>final<|message|>")

trainer = train_on_responses_only(
    trainer,
    **gpt_oss_kwargs,
)

Let's verify masking the instruction part is done! Let's print the 100th row again.

In [ ]:
tokenizer.decode(trainer.train_dataset[100]["input_ids"])

Now let's print the masked out example - you should see only the answer is present:

In [ ]:
tokenizer.decode([tokenizer.pad_token_id if x == -100 else x for x in trainer.train_dataset[100]["labels"]]).replace(tokenizer.pad_token, " ")

In [ ]:
# @title Show current memory stats
gpu_stats = torch.cuda.get_device_properties(0)
start_gpu_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
max_memory = round(gpu_stats.total_memory / 1024 / 1024 / 1024, 3)
print(f"GPU = {gpu_stats.name}. Max memory = {max_memory} GB.")
print(f"{start_gpu_memory} GB of memory reserved.")

GPU = Tesla T4. Max memory = 14.741 GB.
12.811 GB of memory reserved.


Let's train the model! To resume a training run, set `trainer.train(resume_from_checkpoint = True)`

In [ ]:
trainer_stats = trainer.train()

Trainer.tokenizer is now deprecated. You should use Trainer.processing_class instead.
The tokenizer has new special tokens that are also defined in the model configs. The model configs were aligned accordingly. Updated tokens: {'bos_token_id': 199998, 'pad_token_id': 200017}
==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 1,000 | Num Epochs = 1 | Total steps = 30
O^O/ \_/ \    Batch size per device = 1 | Gradient accumulation steps = 4
\        /    Data Parallel GPUs = 1 | Total batch size (1 x 4 x 1) = 4
 "-____-"     Trainable parameters = 3,981,312 of 20,918,738,496 (0.02% trained)
`use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`.


Unsloth: Will smartly offload gradients to save VRAM!


In [ ]:
# @title Show final memory and time stats
used_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
used_memory_for_lora = round(used_memory - start_gpu_memory, 3)
used_percentage = round(used_memory / max_memory * 100, 3)
lora_percentage = round(used_memory_for_lora / max_memory * 100, 3)
print(f"{trainer_stats.metrics['train_runtime']} seconds used for training.")
print(
    f"{round(trainer_stats.metrics['train_runtime']/60, 2)} minutes used for training."
)
print(f"Peak reserved memory = {used_memory} GB.")
print(f"Peak reserved memory for training = {used_memory_for_lora} GB.")
print(f"Peak reserved memory % of max memory = {used_percentage} %.")
print(f"Peak reserved memory for training % of max memory = {lora_percentage} %.")

645.6936 seconds used for training.
10.76 minutes used for training.
Peak reserved memory = 12.975 GB.
Peak reserved memory for training = 0.164 GB.
Peak reserved memory % of max memory = 88.02 %.
Peak reserved memory for training % of max memory = 1.113 %.


<a name="Inference"></a>
### Inference
Let's run the model! You can change the instruction and input - leave the output blank!

In [ ]:
messages = [
    {"role": "system", "content": "reasoning language: French\n\nYou are a helpful assistant that can solve mathematical problems."},
    {"role": "user", "content": "Solve x^5 + 3x^4 - 10 = 3."},
]
inputs = tokenizer.apply_chat_template(
    messages,
    add_generation_prompt = True,
    return_tensors = "pt",
    return_dict = True,
    reasoning_effort = "medium",
).to("cuda")
from transformers import TextStreamer
_ = model.generate(**inputs, max_new_tokens = 64, streamer = TextStreamer(tokenizer))

<|start|>system<|message|>You are ChatGPT, a large language model trained by OpenAI.
Knowledge cutoff: 2024-06
Current date: 2025-08-13

Reasoning: medium

# Valid channels: analysis, commentary, final. Channel must be included for every message.
Calls to these tools must go to the commentary channel: 'functions'.<|end|><|start|>developer<|message|># Instructions

reasoning language: French

You are a helpful assistant that can solve mathematical problems.<|end|><|start|>user<|message|>Solve x^5 + 3x^4 - 10 = 3.<|end|><|start|>assistant<|channel|>analysis<|message|>The equation is \(x^5 + 3x^4 - 10 = 3\), or \(x^5 + 3x^4 - 13 = 0\). So we need to find the roots of \(x^5 + 3x^4 - 13


<a name="Save"></a>
### Saving, loading finetuned models
To save the final model as LoRA adapters, either use Huggingface's `push_to_hub` for an online save or `save_pretrained` for a local save.

**[NOTE]** Currently finetunes can only be loaded via Unsloth in the meantime - we're working on vLLM and GGUF exporting!

In [ ]:
model.save_pretrained("finetuned_model")
# model.push_to_hub("hf_username/finetuned_model", token = "hf_...") # Save to HF

To run the finetuned model, you can do the below after setting `if False` to `if True` in a new instance.

In [ ]:
if True:
    from unsloth import FastLanguageModel
    model, tokenizer = FastLanguageModel.from_pretrained(
        model_name = "finetuned_model", # YOUR MODEL YOU USED FOR TRAINING
        max_seq_length = 1024,
        dtype = None,
        load_in_4bit = True,
    )

messages = [
    {"role": "system", "content": "reasoning language: English\n\nYou are a helpful assistant that can solve mathematical problems."},
    {"role": "user", "content": "Solve x^5 + 3x^4 - 10 = 3."},
]
inputs = tokenizer.apply_chat_template(
    messages,
    add_generation_prompt = True,
    return_tensors = "pt",
    return_dict = True,
    reasoning_effort = "high",
).to("cuda")
from transformers import TextStreamer
_ = model.generate(**inputs, max_new_tokens = 64, streamer = TextStreamer(tokenizer))

<|start|>system<|message|>You are ChatGPT, a large language model trained by OpenAI.
Knowledge cutoff: 2024-06
Current date: 2025-08-13

Reasoning: high

# Valid channels: analysis, commentary, final. Channel must be included for every message.
Calls to these tools must go to the commentary channel: 'functions'.<|end|><|start|>developer<|message|># Instructions

reasoning language: French

You are a helpful assistant that can solve mathematical problems.<|end|><|start|>user<|message|>Solve x^5 + 3x^4 - 10 = 3.<|end|><|start|>assistant<|channel|>analysis<|message|>We need to solve the equation for x. The equation: x^5 + 3x^4 - 10 = 3. So bring 3 to left side: x^5 + 3x^4 -10 -3 = 0 → x^5 + 3x^


### Saving to float16 for VLLM or mxfp4

We also support saving to `float16` or `mxfp4` directly. Select `merged_16bit` for float16. Use `push_to_hub_merged` to upload to your Hugging Face account! You can go to https://huggingface.co/settings/tokens for your personal tokens.

In [ ]:
# Merge and push to hub in mxfp4 4bit format
if False:
    model.save_pretrained_merged("finetuned_model", tokenizer, save_method = "mxfp4")
if False: model.push_to_hub_merged("repo_id/repo_name", tokenizer, token = "hf...", save_method = "mxfp4")

# Merge and push to hub in 16bit
if False:
    model.save_pretrained_merged("finetuned_model", tokenizer, save_method = "merged_16bit")
if False: # Pushing to HF Hub
    model.push_to_hub_merged("hf/gpt-oss-finetune", tokenizer, save_method = "merged_16bit", token = "")

And we're done! If you have any questions on Unsloth, we have a [Discord](https://discord.gg/unsloth) channel! If you find any bugs or want to keep updated with the latest LLM stuff, or need help, join projects etc, feel free to join our Discord!

Some other links:
1. Train your own reasoning model - Llama GRPO notebook [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.1_(8B)-GRPO.ipynb)
2. Saving finetunes to Ollama. [Free notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)
3. Llama 3.2 Vision finetuning - Radiography use case. [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.2_(11B)-Vision.ipynb)
6. See notebooks for DPO, ORPO, Continued pretraining, conversational finetuning and more on our [documentation](https://docs.unsloth.ai/get-started/unsloth-notebooks)!

<div class="align-center">
  <a href="https://unsloth.ai"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
  <a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord.png" width="145"></a>
  <a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a>

  Join Discord if you need help + ⭐️ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐️
</div>

  This notebook and all Unsloth notebooks are licensed [LGPL-3.0](https://github.com/unslothai/notebooks?tab=LGPL-3.0-1-ov-file#readme).


# Task
Finetune an open-source language model using the Unsloth library on a free Tesla T4 Google Colab instance, covering the entire process from installing dependencies, loading a base model (e.g., "unsloth/gpt-oss-20b"), preparing custom data (using "HuggingFaceH4/Multilingual-Thinking" as an example), configuring LoRA adapters, training the model, running inference, and saving the finetuned model, while highlighting the ease of use and computational benefits.

## Verify Colab Environment

### Subtask:
Confirm that you are running in a Google Colab environment, ideally on a free Tesla T4 GPU instance, as specified in the original notebook.


To confirm the Colab environment:

1.  **Check the Colab runtime type**: Click on "Runtime" in the Colab menu, then select "Change runtime type." Ensure that the "Hardware accelerator" is set to "T4 GPU" or similar GPU type, and then click "Save."

2.  **Confirm the GPU allocated**: Execute the code cell below (cell id `IGT9FDxxnox0`) that displays the GPU information to verify that a Tesla T4 GPU (or equivalent) has been successfully allocated. Pay attention to the 'GPU' and 'Max memory' output.

2.  **Confirm the GPU allocated**: Execute the code cell below (cell id `IGT9FDxxnox0`) that displays the GPU information to verify that a Tesla T4 GPU (or equivalent) has been successfully allocated. Pay attention to the 'GPU' and 'Max memory' output.

2.  **Confirm the GPU allocated**: Execute the code cell below (cell id `IGT9FDxxnox0`) that displays the GPU information to verify that a Tesla T4 GPU (or equivalent) has been successfully allocated. Pay attention to the 'GPU' and 'Max memory' output.


2.  **Confirm the GPU allocated**: Execute the code cell below (cell id `IGT9FDxxnox0`) that displays the GPU information to verify that a Tesla T4 GPU (or equivalent) has been successfully allocated. Pay attention to the 'GPU' and 'Max memory' output.


2.  **Confirm the GPU allocated**: Execute the code cell below (cell id `IGT9FDxxnox0`) that displays the GPU information to verify that a Tesla T4 GPU (or equivalent) has been successfully allocated. Pay attention to the 'GPU' and 'Max memory' output.

2.  **Confirm the GPU allocated**: Execute the code cell below (cell id `IGT9FDxxnox0`) that displays the GPU information to verify that a Tesla T4 GPU (or equivalent) has been successfully allocated. Pay attention to the 'GPU' and 'Max memory' output.

2.  **Confirm the GPU allocated**: Execute the code cell below (cell id `IGT9FDxxnox0`) that displays the GPU information to verify that a Tesla T4 GPU (or equivalent) has been successfully allocated. Pay attention to the 'GPU' and 'Max memory' output.

2.  **Confirm the GPU allocated**: Execute the code cell below (cell id `IGT9FDxxnox0`) that displays the GPU information to verify that a Tesla T4 GPU (or equivalent) has been successfully allocated. Pay attention to the 'GPU' and 'Max memory' output.

3. Once confirmed, execute the next code cell to proceed with the notebook.

3. Once confirmed, execute the next code cell to proceed with the notebook.

3. Once confirmed, execute the next code cell to proceed with the notebook.

3. Once confirmed, execute the next code cell to proceed with the notebook.

3. Once confirmed, execute the next code cell to proceed with the notebook.

3. Once confirmed, execute the next code cell to proceed with the notebook.

## Install Unsloth and Dependencies

### Subtask:
Install the Unsloth library and all necessary dependencies for model training and inference. This step uses `uv pip install` commands, which are optimized for Colab.


**Reasoning**:
The subtask requires installing Unsloth and its dependencies. The provided instruction explicitly asks to execute the code cell with ID `KZhsxIbZnoxq`, which contains the necessary installation commands. Therefore, I will execute this existing code block.



In [1]:
%%capture
import os, importlib.util
!pip install --upgrade -qqq uv
if importlib.util.find_spec("torch") is None or "COLAB_" in "".join(os.environ.keys()):
    try: import numpy, PIL; get_numpy = f"numpy=={numpy.__version__}"; get_pil = f"pillow=={PIL.__version__}"
    except: get_numpy = "numpy"; get_pil = "pillow"
    !uv pip install -qqq \
        "torch>=2.8.0" "triton>=3.4.0" {get_numpy} {get_pil} torchvision bitsandbytes "transformers==4.56.2" \
        "unsloth_zoo[base] @ git+https://github.com/unslothai/unsloth-zoo" \
        "unsloth[base] @ git+https://github.com/unslothai/unsloth" \
        git+https://github.com/triton-lang/triton.git@0add68262ab0a2e33b84524346cb27cbb2787356#subdirectory=python/triton_kernels
elif importlib.util.find_spec("unsloth") is None:
    !uv pip install -qqq unsloth
!uv pip install --upgrade --no-deps transformers==4.56.2 tokenizers trl==0.22.2 unsloth unsloth_zoo

## Load Base Open Source Model

### Subtask:
Initialize and load your chosen base open-source language model using `FastLanguageModel.from_pretrained`.


**Reasoning**:
The subtask requires loading the base model and tokenizer by executing the specified code cell.



In [2]:
from unsloth import FastLanguageModel
import torch
max_seq_length = 1024
dtype = None

# 4bit pre quantized models we support for 4x faster downloading + no OOMs.
fourbit_models = [
    "unsloth/gpt-oss-20b-unsloth-bnb-4bit", # 20B model using bitsandbytes 4bit quantization
    "unsloth/gpt-oss-120b-unsloth-bnb-4bit",
    "unsloth/gpt-oss-20b", # 20B model using MXFP4 format
    "unsloth/gpt-oss-120b",
] # More models at https://huggingface.co/unsloth

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "unsloth/gpt-oss-20b",
    dtype = dtype, # None for auto detection
    max_seq_length = max_seq_length, # Choose any for long context!
    load_in_4bit = True,  # 4 bit quantization to reduce memory
    full_finetuning = False, # [NEW!] We have full finetuning now!
    # token = "hf_...", # use one if using gated models
)

🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.


/usr/local/lib/python3.12/dist-packages/torchao/quantization/quant_api.py:2525: SyntaxWarning: invalid escape sequence '\.'
  * regex for parameter names, must start with `re:`, e.g. `re:language\.layers\..+\.q_proj.weight`.


🦥 Unsloth Zoo will now patch everything to make training faster!
==((====))==  Unsloth 2025.12.9: Fast Gpt_Oss patching. Transformers: 4.56.2.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.9.0+cu126. CUDA: 7.5. CUDA Toolkit: 12.6. Triton: 3.5.0
\        /    Bfloat16 = FALSE. FA [Xformers = None. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!
Unsloth: Using float16 precision for gpt_oss won't work! Using float32.


model.safetensors.index.json: 0.00B [00:00, ?B/s]

model-00001-of-00004.safetensors:   0%|          | 0.00/4.00G [00:00<?, ?B/s]

model-00002-of-00004.safetensors:   0%|          | 0.00/4.00G [00:00<?, ?B/s]

model-00003-of-00004.safetensors:   0%|          | 0.00/3.37G [00:00<?, ?B/s]

model-00004-of-00004.safetensors:   0%|          | 0.00/1.16G [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/4 [00:00<?, ?it/s]

KeyboardInterrupt: 

# Task
It looks like the previous execution of "Load Base Open Source Model" was interrupted. Let's try running that cell again to ensure the model and tokenizer are loaded correctly before configuring LoRA adapters.

## Load Base Open Source Model

### Subtask:
Initialize and load your chosen base open-source language model using `FastLanguageModel.from_pretrained`.

### Reasoning:
The subtask requires loading the base model and tokenizer by executing the specified code cell.

### Execute:
The model was not loaded successfully in the previous attempt due to a KeyboardInterrupt. We need to re-execute the cell to load the model before proceeding.

```python
from unsloth import FastLanguageModel
import torch
max_seq_length = 1024
dtype = None

# 4bit pre quantized models we support for 4x faster downloading + no OOMs.
fourbit_models = [
    "unsloth/gpt-oss-20b-unsloth-bnb-4bit", # 20B model using bitsandbytes 4bit quantization
    "unsloth/gpt-oss-120b-unsloth-bnb-4bit",
    "unsloth/gpt-oss-20b", # 20B model using MXFP4 format
    "unsloth/gpt-oss-120b",
] # More models at https://huggingface.co/unsloth

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "unsloth/gpt-oss-20b",
    dtype = dtype, # None for auto detection
    max_seq_length = max_seq_length, # Choose any for long context!
    load_in_4bit = True,  # 4 bit quantization to reduce memory
    full_finetuning = False, # [NEW] We have full finetuning now!
    # token = "hf_...", # use one if using gated models
)
```

## Load Base Open Source Model

### Subtask:
Initialize and load your chosen base open-source language model using `FastLanguageModel.from_pretrained`.


**Reasoning**:
The subtask requires loading the base model and tokenizer, so I will execute the corresponding code cell to perform this action.



In [3]:
from unsloth import FastLanguageModel
import torch
max_seq_length = 1024
dtype = None

# 4bit pre quantized models we support for 4x faster downloading + no OOMs.
fourbit_models = [
    "unsloth/gpt-oss-20b-unsloth-bnb-4bit", # 20B model using bitsandbytes 4bit quantization
    "unsloth/gpt-oss-120b-unsloth-bnb-4bit",
    "unsloth/gpt-oss-20b", # 20B model using MXFP4 format
    "unsloth/gpt-oss-120b",
] # More models at https://huggingface.co/unsloth

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "unsloth/gpt-oss-20b",
    dtype = dtype, # None for auto detection
    max_seq_length = max_seq_length, # Choose any for long context!
    load_in_4bit = True,  # 4 bit quantization to reduce memory
    full_finetuning = False, # [NEW!] We have full finetuning now!
    # token = "hf_...", # use one if using gated models
)

==((====))==  Unsloth 2025.12.9: Fast Gpt_Oss patching. Transformers: 4.56.2.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.9.0+cu126. CUDA: 7.5. CUDA Toolkit: 12.6. Triton: 3.5.0
\        /    Bfloat16 = FALSE. FA [Xformers = None. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!
Unsloth: Using float16 precision for gpt_oss won't work! Using float32.


ValueError: Some modules are dispatched on the CPU or the disk. Make sure you have enough GPU RAM to fit the quantized model. If you want to dispatch the model on the CPU or the disk while keeping these modules in 32-bit, you need to set `llm_int8_enable_fp32_cpu_offload=True` and pass a custom `device_map` to `from_pretrained`. Check https://huggingface.co/docs/transformers/main/en/main_classes/quantization#offload-between-cpu-and-gpu for more details. 

# Task
The previous attempt to load the base model encountered a `ValueError` indicating insufficient GPU RAM. To address this, we will add `device_map='auto'` to the `FastLanguageModel.from_pretrained` call. This setting allows the model to automatically distribute its layers across available devices (GPU and CPU) to manage memory efficiently and prevent out-of-memory errors.

## Load Base Open Source Model

### Subtask:
Initialize and load your chosen base open-source language model using `FastLanguageModel.from_pretrained` with `device_map='auto'` to mitigate memory errors.

### Reasoning:
The previous execution failed due to a `ValueError` related to GPU memory. Adding `device_map='auto'` as suggested by the initial plan will allow the model to be loaded efficiently across available resources, preventing memory issues during the loading process.

### Execute:
Modify the `from_pretrained` function to include `device_map='auto'`.

```python
from unsloth import FastLanguageModel
import torch
max_seq_length = 1024
dtype = None

# 4bit pre quantized models we support for 4x faster downloading + no OOMs.
fourbit_models = [
    "unsloth/gpt-oss-20b-unsloth-bnb-4bit", # 20B model using bitsandbytes 4bit quantization
    "unsloth/gpt-oss-120b-unsloth-bnb-4bit",
    "unsloth/gpt-oss-20b", # 20B model using MXFP4 format
    "unsloth/gpt-oss-120b",
] # More models at https://huggingface.co/unsloth

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "unsloth/gpt-oss-20b",
    dtype = dtype, # None for auto detection
    max_seq_length = max_seq_length, # Choose any for long context!
    load_in_4bit = True,  # 4 bit quantization to reduce memory
    full_finetuning = False, # [NEW!] We have full finetuning now!
    # token = "hf_...", # use one if using gated models
    device_map = "auto", # Added to mitigate memory errors
)
```

## Load Base Open Source Model

### Subtask:
Initialize and load your chosen base open-source language model using `FastLanguageModel.from_pretrained` with `device_map='auto'` to mitigate memory errors.


**Reasoning**:
The subtask requires loading the base model and tokenizer, and the previous attempt failed due to memory issues. By adding `device_map='auto'` to the `FastLanguageModel.from_pretrained` call, we instruct Hugging Face to automatically distribute the model across available devices, which should mitigate memory errors and allow the model to load successfully.



In [4]:
from unsloth import FastLanguageModel
import torch
max_seq_length = 1024
dtype = None

# 4bit pre quantized models we support for 4x faster downloading + no OOMs.
fourbit_models = [
    "unsloth/gpt-oss-20b-unsloth-bnb-4bit", # 20B model using bitsandbytes 4bit quantization
    "unsloth/gpt-oss-120b-unsloth-bnb-4bit",
    "unsloth/gpt-oss-20b", # 20B model using MXFP4 format
    "unsloth/gpt-oss-120b",
] # More models at https://huggingface.co/unsloth

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "unsloth/gpt-oss-20b",
    dtype = dtype, # None for auto detection
    max_seq_length = max_seq_length, # Choose any for long context!
    load_in_4bit = True,  # 4 bit quantization to reduce memory
    full_finetuning = False, # [NEW!] We have full finetuning now!
    device_map = "auto", # Added to handle memory distribution automatically
    # token = "hf_...", # use one if using gated models
)

==((====))==  Unsloth 2025.12.9: Fast Gpt_Oss patching. Transformers: 4.56.2.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.9.0+cu126. CUDA: 7.5. CUDA Toolkit: 12.6. Triton: 3.5.0
\        /    Bfloat16 = FALSE. FA [Xformers = None. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!
Unsloth: Using float16 precision for gpt_oss won't work! Using float32.


ValueError: Some modules are dispatched on the CPU or the disk. Make sure you have enough GPU RAM to fit the quantized model. If you want to dispatch the model on the CPU or the disk while keeping these modules in 32-bit, you need to set `llm_int8_enable_fp32_cpu_offload=True` and pass a custom `device_map` to `from_pretrained`. Check https://huggingface.co/docs/transformers/main/en/main_classes/quantization#offload-between-cpu-and-gpu for more details. 

**Reasoning**:
The previous attempts to load the model failed because adding `device_map='auto'` conflicted with Unsloth's internal 4-bit quantization memory management, leading to a `ValueError`. The original notebook's code for loading the model did not include `device_map='auto'` and executed successfully. Therefore, the fix is to revert to the original `FastLanguageModel.from_pretrained` call, allowing Unsloth to handle device mapping automatically with `load_in_4bit=True`.



In [5]:
from unsloth import FastLanguageModel
import torch
max_seq_length = 1024
dtype = None

# 4bit pre quantized models we support for 4x faster downloading + no OOMs.
fourbit_models = [
    "unsloth/gpt-oss-20b-unsloth-bnb-4bit", # 20B model using bitsandbytes 4bit quantization
    "unsloth/gpt-oss-120b-unsloth-bnb-4bit",
    "unsloth/gpt-oss-20b", # 20B model using MXFP4 format
    "unsloth/gpt-oss-120b",
] # More models at https://huggingface.co/unsloth

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "unsloth/gpt-oss-20b",
    dtype = dtype, # None for auto detection
    max_seq_length = max_seq_length, # Choose any for long context!
    load_in_4bit = True,  # 4 bit quantization to reduce memory
    full_finetuning = False, # [NEW!] We have full finetuning now!
    # token = "hf_...", # use one if using gated models
)

==((====))==  Unsloth 2025.12.9: Fast Gpt_Oss patching. Transformers: 4.56.2.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.9.0+cu126. CUDA: 7.5. CUDA Toolkit: 12.6. Triton: 3.5.0
\        /    Bfloat16 = FALSE. FA [Xformers = None. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!
Unsloth: Using float16 precision for gpt_oss won't work! Using float32.


ValueError: Some modules are dispatched on the CPU or the disk. Make sure you have enough GPU RAM to fit the quantized model. If you want to dispatch the model on the CPU or the disk while keeping these modules in 32-bit, you need to set `llm_int8_enable_fp32_cpu_offload=True` and pass a custom `device_map` to `from_pretrained`. Check https://huggingface.co/docs/transformers/main/en/main_classes/quantization#offload-between-cpu-and-gpu for more details. 

**Reasoning**:
The previous attempts to load the base model failed with a `ValueError` indicating insufficient GPU RAM, even with `load_in_4bit=True`. The traceback explicitly suggests setting `llm_int8_enable_fp32_cpu_offload=True` and passing a custom `device_map` if modules are dispatched to CPU or disk. To directly address this, I will add both `device_map="auto"` and `llm_int8_enable_fp32_cpu_offload=True` to the `from_pretrained` call to allow modules to offload to CPU while keeping them in 32-bit if necessary, which might resolve the memory allocation issue.



In [6]:
from unsloth import FastLanguageModel
import torch
max_seq_length = 1024
dtype = None

# 4bit pre quantized models we support for 4x faster downloading + no OOMs.
fourbit_models = [
    "unsloth/gpt-oss-20b-unsloth-bnb-4bit", # 20B model using bitsandbytes 4bit quantization
    "unsloth/gpt-oss-120b-unsloth-bnb-4bit",
    "unsloth/gpt-oss-20b", # 20B model using MXFP4 format
    "unsloth/gpt-oss-120b",
] # More models at https://huggingface.co/unsloth

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "unsloth/gpt-oss-20b",
    dtype = dtype, # None for auto detection
    max_seq_length = max_seq_length, # Choose any for long context!
    load_in_4bit = True,  # 4 bit quantization to reduce memory
    full_finetuning = False, # [NEW!] We have full finetuning now!
    # token = "hf_...", # use one if using gated models
    device_map = "auto", # Added to handle memory distribution automatically
    llm_int8_enable_fp32_cpu_offload=True, # Added as per error message recommendation
)

==((====))==  Unsloth 2025.12.9: Fast Gpt_Oss patching. Transformers: 4.56.2.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.9.0+cu126. CUDA: 7.5. CUDA Toolkit: 12.6. Triton: 3.5.0
\        /    Bfloat16 = FALSE. FA [Xformers = None. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!
Unsloth: Using float16 precision for gpt_oss won't work! Using float32.


TypeError: GptOssForCausalLM.__init__() got an unexpected keyword argument 'llm_int8_enable_fp32_cpu_offload'

**Reasoning**:
The previous attempts to load the base model failed due to a `TypeError` because `llm_int8_enable_fp32_cpu_offload` is an unexpected keyword argument for `GptOssForCausalLM` when loaded via `unsloth`. The original notebook code, which successfully loaded the model, did not include `device_map` or `llm_int8_enable_fp32_cpu_offload`. Unsloth's `FastLanguageModel.from_pretrained` with `load_in_4bit=True` is designed to handle memory management automatically, so these additional arguments are unnecessary and cause conflicts.



In [7]:
from unsloth import FastLanguageModel
import torch
max_seq_length = 1024
dtype = None

# 4bit pre quantized models we support for 4x faster downloading + no OOMs.
fourbit_models = [
    "unsloth/gpt-oss-20b-unsloth-bnb-4bit", # 20B model using bitsandbytes 4bit quantization
    "unsloth/gpt-oss-120b-unsloth-bnb-4bit",
    "unsloth/gpt-oss-20b", # 20B model using MXFP4 format
    "unsloth/gpt-oss-120b",
] # More models at https://huggingface.co/unsloth

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "unsloth/gpt-oss-20b",
    dtype = dtype, # None for auto detection
    max_seq_length = max_seq_length, # Choose any for long context!
    load_in_4bit = True,  # 4 bit quantization to reduce memory
    full_finetuning = False, # [NEW!] We have full finetuning now!
    # token = "hf_...", # use one if using gated models
)

==((====))==  Unsloth 2025.12.9: Fast Gpt_Oss patching. Transformers: 4.56.2.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.9.0+cu126. CUDA: 7.5. CUDA Toolkit: 12.6. Triton: 3.5.0
\        /    Bfloat16 = FALSE. FA [Xformers = None. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!
Unsloth: Using float16 precision for gpt_oss won't work! Using float32.


ValueError: Some modules are dispatched on the CPU or the disk. Make sure you have enough GPU RAM to fit the quantized model. If you want to dispatch the model on the CPU or the disk while keeping these modules in 32-bit, you need to set `llm_int8_enable_fp32_cpu_offload=True` and pass a custom `device_map` to `from_pretrained`. Check https://huggingface.co/docs/transformers/main/en/main_classes/quantization#offload-between-cpu-and-gpu for more details. 

**Reasoning**:
The previous step failed with a "TypeError" because `llm_int8_enable_fp32_cpu_offload` is not a recognized argument for `FastLanguageModel.from_pretrained` or the underlying `GptOssForCausalLM`. To resolve this `TypeError`, I must remove the `llm_int8_enable_fp32_cpu_offload` argument. I will keep `device_map="auto"` as it is intended to help with memory management.



In [8]:
from unsloth import FastLanguageModel
import torch
max_seq_length = 1024
dtype = None

# 4bit pre quantized models we support for 4x faster downloading + no OOMs.
fourbit_models = [
    "unsloth/gpt-oss-20b-unsloth-bnb-4bit", # 20B model using bitsandbytes 4bit quantization
    "unsloth/gpt-oss-120b-unsloth-bnb-4bit",
    "unsloth/gpt-oss-20b", # 20B model using MXFP4 format
    "unsloth/gpt-oss-120b",
] # More models at https://huggingface.co/unsloth

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "unsloth/gpt-oss-20b",
    dtype = dtype, # None for auto detection
    max_seq_length = max_seq_length, # Choose any for long context!
    load_in_4bit = True,  # 4 bit quantization to reduce memory
    full_finetuning = False, # [NEW!] We have full finetuning now!
    # token = "hf_...", # use one if using gated models
    device_map = "auto", # Keep device_map="auto" to aid with memory distribution
)

==((====))==  Unsloth 2025.12.9: Fast Gpt_Oss patching. Transformers: 4.56.2.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.9.0+cu126. CUDA: 7.5. CUDA Toolkit: 12.6. Triton: 3.5.0
\        /    Bfloat16 = FALSE. FA [Xformers = None. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!
Unsloth: Using float16 precision for gpt_oss won't work! Using float32.


ValueError: Some modules are dispatched on the CPU or the disk. Make sure you have enough GPU RAM to fit the quantized model. If you want to dispatch the model on the CPU or the disk while keeping these modules in 32-bit, you need to set `llm_int8_enable_fp32_cpu_offload=True` and pass a custom `device_map` to `from_pretrained`. Check https://huggingface.co/docs/transformers/main/en/main_classes/quantization#offload-between-cpu-and-gpu for more details. 

**Reasoning**:
The model loading failed repeatedly with a `ValueError` recommending `llm_int8_enable_fp32_cpu_offload=True` and a custom `device_map`. To properly pass `llm_int8_enable_fp32_cpu_offload`, it needs to be set within a `BitsAndBytesConfig` object, which is then passed to `FastLanguageModel.from_pretrained` via the `quantization_config` argument, along with `device_map='auto'` to handle memory distribution.



In [9]:
from unsloth import FastLanguageModel
import torch
from transformers import BitsAndBytesConfig

max_seq_length = 1024
dtype = None

# 4bit pre quantized models we support for 4x faster downloading + no OOMs.
fourbit_models = [
    "unsloth/gpt-oss-20b-unsloth-bnb-4bit", # 20B model using bitsandbytes 4bit quantization
    "unsloth/gpt-oss-120b-unsloth-bnb-4bit",
    "unsloth/gpt-oss-20b", # 20B model using MXFP4 format
    "unsloth/gpt-oss-120b",
] # More models at https://huggingface.co/unsloth

# Configure BitsAndBytes for 4-bit loading with CPU offload for 32-bit modules
bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch.float16,
    bnb_4bit_use_double_quant=True,
    llm_int8_enable_fp32_cpu_offload=True, # Explicitly enable 32-bit CPU offload
)

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "unsloth/gpt-oss-20b",
    dtype = dtype, # None for auto detection
    max_seq_length = max_seq_length, # Choose any for long context!
    quantization_config = bnb_config, # Pass the custom quantization config
    full_finetuning = False, # [NEW!] We have full finetuning now!
    device_map = "auto", # Keep device_map="auto" to aid with memory distribution
    # token = "hf_...", # use one if using gated models
)

==((====))==  Unsloth 2025.12.9: Fast Gpt_Oss patching. Transformers: 4.56.2.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.9.0+cu126. CUDA: 7.5. CUDA Toolkit: 12.6. Triton: 3.5.0
\        /    Bfloat16 = FALSE. FA [Xformers = None. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!
Unsloth: Using float16 precision for gpt_oss won't work! Using float32.


Loading checkpoint shards:   0%|          | 0/4 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/165 [00:00<?, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

tokenizer.json:   0%|          | 0.00/27.9M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/446 [00:00<?, ?B/s]

chat_template.jinja: 0.00B [00:00, ?B/s]

## Configure LoRA Adapters

### Subtask:
Set up LoRA (Low-Rank Adaptation) adapters on your loaded model using `FastLanguageModel.get_peft_model`. This step will optimize memory usage on the Colab GPU for efficient finetuning.


**Reasoning**:
The subtask requires setting up LoRA adapters on the loaded model. The provided instructions indicate that this is achieved by executing the code cell with ID `f_LK81NRnVgg`, which contains the `FastLanguageModel.get_peft_model` call. I will execute this cell to configure the model for parameter-efficient finetuning.



In [10]:
model = FastLanguageModel.get_peft_model(
    model,
    r = 8, # Choose any number > 0 ! Suggested 8, 16, 32, 64, 128
    target_modules = ["q_proj", "k_proj", "v_proj", "o_proj",
                      "gate_proj", "up_proj", "down_proj"],
    lora_alpha = 16,
    lora_dropout = 0, # Supports any, but = 0 is optimized
    bias = "none",    # Supports any, but = "none" is optimized
    # [NEW] "unsloth" uses 30% less VRAM, fits 2x larger batch sizes!
    use_gradient_checkpointing = "unsloth", # True or "unsloth" for very long context
    random_state = 3407,
    use_rslora = False,  # We support rank stabilized LoRA
    loftq_config = None, # And LoftQ
)

Unsloth: Making `model.base_model.model.model` require gradients


## Prepare Custom Data

### Subtask:
Load and preprocess your custom dataset for finetuning. This involves defining a formatting function and applying it to the dataset using the loaded tokenizer.


**Reasoning**:
The subtask requires loading and preprocessing the custom dataset for finetuning, which involves defining a formatting function and applying it to the dataset. The existing code cell `62QfuPXBnVgi` in the notebook already performs these exact steps, so I will execute it.



In [11]:
def formatting_prompts_func(examples):
    convos = examples["messages"]
    texts = [tokenizer.apply_chat_template(convo, tokenize = False, add_generation_prompt = False) for convo in convos]
    return { "text" : texts, }

from datasets import load_dataset

def formatting_prompts_func(examples):
    convos = examples["messages"]
    texts = [
        tokenizer.apply_chat_template(
            convo,
            tokenize=False,
            add_generation_prompt=False
        )
        for convo in convos
    ]
    return {"text": texts}

dataset = load_dataset(
    "json",
    data_files={"train": "nz_sovereign_train_v1.jsonl"},
    split="train",
)

dataset = dataset.map(formatting_prompts_func, batched=True, remove_columns=dataset.column_names)
dataset

Generating train split: 0 examples [00:00, ? examples/s]

Map:   0%|          | 0/184 [00:00<?, ? examples/s]

Dataset({
    features: ['text'],
    num_rows: 184
})

## Train the Model

### Subtask:
Configure and initiate the training process using `trl.SFTTrainer`.


**Reasoning**:
The subtask requires configuring and initiating the training process. The first step is to initialize the `SFTTrainer` with the model, tokenizer, and dataset, along with specified training arguments. The existing code cell `O-XZLeLYnVgk` already performs this action.



In [12]:
from trl import SFTConfig, SFTTrainer
trainer = SFTTrainer(
    model = model,
    tokenizer = tokenizer,
    train_dataset = dataset,
    args = SFTConfig(
        per_device_train_batch_size = 1,
        gradient_accumulation_steps = 4,
        warmup_steps = 5,
        num_train_epochs = 1, # Set this for 1 full training run.
        max_steps = 30,
        learning_rate = 2e-4,
        logging_steps = 1,
        optim = "adamw_8bit",
        weight_decay = 0.001,
        lr_scheduler_type = "linear",
        seed = 3407,
        output_dir = "outputs",
        report_to = "none", # Use TrackIO/WandB etc
    ),
)

Unsloth: Switching to float32 training since model cannot work with float16


Unsloth: Tokenizing ["text"] (num_proc=4):   0%|          | 0/184 [00:00<?, ? examples/s]

**Reasoning**:
The next step in the subtask is to apply the `train_on_responses_only` utility from Unsloth to the `trainer` object to ensure the model learns only from assistant responses. The code cell `bY_0OUponoxz` performs this action.



In [13]:
from unsloth.chat_templates import train_on_responses_only

gpt_oss_kwargs = dict(instruction_part = "<|start|>user<|message|>", response_part="<|start|>assistant<|channel|>final<|message|>")

trainer = train_on_responses_only(
    trainer,
    **gpt_oss_kwargs,
)

Map (num_proc=6):   0%|          | 0/184 [00:00<?, ? examples/s]

ZeroDivisionError: Unsloth: All labels in your dataset are -100. Training losses will be all 0.
For example, are you sure you used `train_on_responses_only` correctly?
Or did you mask our tokens incorrectly? Maybe this is intended?
Maybe you're using a Llama chat template on a non Llama model for example?If you used `train_on_responses_only`, confirm your user and assistant parts are correct!

**Reasoning**:
The previous step failed with a `ZeroDivisionError` because `train_on_responses_only` could not find the specified `response_part` in the dataset, resulting in all labels being masked as -100. This indicates a mismatch between the `response_part` defined in `gpt_oss_kwargs` and the actual chat template used in the training data. Observed inference examples in the notebook consistently use the `analysis` channel for assistant responses, not `final`. Therefore, I will update the `response_part` to target the `analysis` channel to correctly identify the start of assistant responses for training.



In [14]:
from unsloth.chat_templates import train_on_responses_only

gpt_oss_kwargs = dict(instruction_part = "<|start|>user<|message|>", response_part="<|start|>assistant<|channel|>analysis<|message|>")

trainer = train_on_responses_only(
    trainer,
    **gpt_oss_kwargs,
)

Map (num_proc=6):   0%|          | 0/184 [00:00<?, ? examples/s]

AttributeError: 'list' object has no attribute 'tolist'

**Reasoning**:
The previous step failed with an `AttributeError: 'list' object has no attribute 'tolist'` within `unsloth_zoo/dataset_utils.py`, indicating that the `examples["labels"]` object, when processed in batches by `datasets.map` with multiprocessing, was a Python list, not a tensor or array that has a `.tolist()` method. To address this library-level issue, I will pass `num_proc=1` to the `train_on_responses_only` function, which will force the `datasets.map` operation to run in a single process, potentially avoiding the data type conversion issue that leads to this error.



In [15]:
from unsloth.chat_templates import train_on_responses_only

gpt_oss_kwargs = dict(instruction_part = "<|start|>user<|message|>", response_part="<|start|>assistant<|channel|>analysis<|message|>")

trainer = train_on_responses_only(
    trainer,
    **gpt_oss_kwargs,
    num_proc = 1, # Force single process to avoid .tolist() error on lists
)

Map (num_proc=1):   0%|          | 0/184 [00:00<?, ? examples/s]

AttributeError: 'list' object has no attribute 'tolist'

**Reasoning**:
The previous attempt to apply `train_on_responses_only` failed with an `AttributeError: 'list' object has no attribute 'tolist'`. This error occurs because `unsloth_zoo` expects the `labels` column in the dataset to be a `torch.Tensor` or `numpy.ndarray` (which have a `.tolist()` method), but it is receiving a plain Python `list` from the `datasets.Dataset.map` function. To ensure `labels` are provided as `torch.Tensor`s, I will explicitly set the format of the `trainer.train_dataset` to 'torch' before applying `train_on_responses_only`.



In [16]:
from unsloth.chat_templates import train_on_responses_only

gpt_oss_kwargs = dict(instruction_part = "<|start|>user<|message|>", response_part="<|start|>assistant<|channel|>analysis<|message|>")

# Explicitly set the format of the dataset to 'torch' to ensure labels are tensors
trainer.train_dataset.set_format(type="torch", columns=["input_ids", "labels"])

trainer = train_on_responses_only(
    trainer,
    **gpt_oss_kwargs,
    num_proc = 1, # Keep num_proc=1 to avoid potential multiprocessing issues
)

Map (num_proc=1):   0%|          | 0/184 [00:00<?, ? examples/s]

AttributeError: 'list' object has no attribute 'tolist'

**Reasoning**:
The previous attempt to apply `train_on_responses_only` failed with an `AttributeError: 'list' object has no attribute 'tolist'` because the `tokenizer` was not explicitly passed to the function, causing it to incorrectly process the dataset and its labels. By explicitly passing the `tokenizer` to `train_on_responses_only`, the function can properly tokenize the dataset and generate the `input_ids` and `labels` in the expected format before applying the masking logic.



In [17]:
from unsloth.chat_templates import train_on_responses_only

gpt_oss_kwargs = dict(instruction_part = "<|start|>user<|message|>", response_part="<|start|>assistant<|channel|>analysis<|message|>")

trainer = train_on_responses_only(
    trainer,
    **gpt_oss_kwargs,
    tokenizer = tokenizer, # Explicitly pass the tokenizer
    num_proc = 1, # Keep num_proc=1 to avoid potential multiprocessing issues
)

Map (num_proc=1):   0%|          | 0/184 [00:00<?, ? examples/s]

AttributeError: 'list' object has no attribute 'tolist'

**Reasoning**:
The `AttributeError: 'list' object has no attribute 'tolist'` occurs because `unsloth_zoo`'s `_train_on_responses_only` function expects `examples["labels"]` to be a NumPy array or PyTorch tensor (which have a `.tolist()` method), but `datasets.map(batched=True)` often passes it as a plain Python list of tensors or lists. To resolve this, I will add a preprocessing step that explicitly maps the `trainer.train_dataset` to flatten the `input_ids` and `labels` and convert them into single NumPy arrays for each batch, ensuring the `.tolist()` method is available.



In [18]:
import numpy as np

def convert_to_numpy_for_unsloth(examples):
    processed_examples = {}

    if "input_ids" in examples:
        flattened_input_ids = [item for sublist in examples["input_ids"] for item in sublist]
        processed_examples["input_ids"] = np.array(flattened_input_ids, dtype=np.int64)

    if "labels" in examples:
        flattened_labels = [item for sublist in examples["labels"] for item in sublist]
        processed_examples["labels"] = np.array(flattened_labels, dtype=np.int64)

    for key, value in examples.items():
        if key not in ["input_ids", "labels"]:
            processed_examples[key] = value

    return processed_examples

# Apply this conversion to the trainer.train_dataset after SFTTrainer initialization
# but before calling train_on_responses_only
trainer.train_dataset = trainer.train_dataset.map(
    convert_to_numpy_for_unsloth,
    batched=True,
    load_from_cache_file=False,
    num_proc=1,
)
print("Trainer dataset converted input_ids and labels to NumPy arrays.")

Map (num_proc=1):   0%|          | 0/184 [00:00<?, ? examples/s]

ArrowInvalid: Column 1 named input_ids expected length 184 but got length 36808

**Reasoning**:
The previous `AttributeError: 'list' object has no attribute 'tolist'` occurred because `unsloth_zoo`'s `_train_on_responses_only` function expected `examples["labels"]` to be a `torch.Tensor` or `numpy.ndarray` for the `.tolist()` method, but it was receiving a plain Python list of lists. To address this, I will explicitly set the format of the `trainer.train_dataset` to 'torch' for the `input_ids` and `labels` columns immediately before calling `train_on_responses_only` to ensure the data is in the expected tensor format.



In [19]:
from unsloth.chat_templates import train_on_responses_only

gpt_oss_kwargs = dict(instruction_part = "<|start|>user<|message|>", response_part="<|start|>assistant<|channel|>analysis<|message|>")

# Ensure the dataset format is set to torch for input_ids and labels
trainer.train_dataset.set_format(type="torch", columns=['input_ids', 'labels'])

trainer = train_on_responses_only(
    trainer,
    **gpt_oss_kwargs,
    tokenizer = tokenizer, # Explicitly pass the tokenizer
    num_proc = 1, # Keep num_proc=1 to avoid potential multiprocessing issues
)

Map (num_proc=1):   0%|          | 0/184 [00:00<?, ? examples/s]

AttributeError: 'list' object has no attribute 'tolist'

**Reasoning**:
The persistent `AttributeError: 'list' object has no attribute 'tolist'` occurs because `unsloth_zoo`'s internal `_train_on_responses_only` function, when operating in batched mode, expects the `labels` for the entire batch to be a single `torch.Tensor` or `numpy.ndarray` that can be converted to a list. However, it seems to receive `labels` as a Python list of individual label tensors/lists for each example in the batch, causing the error. To resolve this, I will explicitly tokenize and pad the dataset to create `input_ids` and `labels` columns as 2D PyTorch tensors *before* calling `train_on_responses_only`. This ensures `_train_on_responses_only` receives the `labels` in the format it expects.



In [20]:
from unsloth.chat_templates import train_on_responses_only
import torch

# First, manually tokenize the dataset to ensure input_ids and labels are present as padded tensors
def tokenize_and_prepare_labels(examples):
    # Tokenize the text column
    tokenized_inputs = tokenizer(
        examples["text"],
        truncation=True,
        max_length=max_seq_length,
        padding="max_length", # Pad to max_seq_length for uniform tensor shape
        return_tensors="pt"
    )

    # The 'labels' are typically a copy of 'input_ids' for language modeling
    # We extract them as lists of lists (for compatibility with datasets.map which expects lists for batched output)
    # and then convert to torch.tensor later or ensure they are properly handled by unsloth
    examples["input_ids"] = tokenized_inputs["input_ids"].tolist()
    examples["labels"] = tokenized_inputs["input_ids"].tolist() # Labels are usually input_ids for casual LM
    examples["attention_mask"] = tokenized_inputs["attention_mask"].tolist()
    return examples

# Apply manual tokenization and padding
tokenized_dataset = trainer.train_dataset.map(
    tokenize_and_prepare_labels,
    batched=True,
    remove_columns=trainer.train_dataset.column_names, # Remove original columns, keep only tokenized ones
    num_proc=1, # Ensure single process to avoid multiprocessing issues during tokenization
    load_from_cache_file=False,
)

# Update the trainer's dataset to the tokenized one
trainer.train_dataset = tokenized_dataset

# Now, apply train_on_responses_only with the already tokenized dataset
gpt_oss_kwargs = dict(instruction_part = "<|start|>user<|message|>", response_part="<|start|>assistant<|channel|>analysis<|message|>")

trainer = train_on_responses_only(
    trainer,
    **gpt_oss_kwargs,
    tokenizer = tokenizer, # Explicitly pass the tokenizer
    num_proc = 1, # Keep num_proc=1 to avoid potential multiprocessing issues
)

print("Dataset tokenized and train_on_responses_only applied.")

Map (num_proc=1):   0%|          | 0/184 [00:00<?, ? examples/s]

Map (num_proc=1):   0%|          | 0/184 [00:00<?, ? examples/s]

Dataset tokenized and train_on_responses_only applied.


**Reasoning**:
The next step is to verify the memory statistics before starting the training to ensure sufficient GPU resources, as per the subtask instructions. The original notebook's cell `IGT9FDxxnox0` provides this functionality.



In [21]:
gpu_stats = torch.cuda.get_device_properties(0)
start_gpu_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
max_memory = round(gpu_stats.total_memory / 1024 / 1024 / 1024, 3)
print(f"GPU = {gpu_stats.name}. Max memory = {max_memory} GB.")
print(f"{start_gpu_memory} GB of memory reserved.")

GPU = Tesla T4. Max memory = 14.741 GB.
12.816 GB of memory reserved.


**Reasoning**:
Now that the memory statistics have been verified and the dataset is properly prepared, the next step is to initiate the training process using `trainer.train()` as per the subtask instructions.



In [22]:
trainer_stats = trainer.train()

==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 184 | Num Epochs = 1 | Total steps = 30
O^O/ \_/ \    Batch size per device = 1 | Gradient accumulation steps = 4
\        /    Data Parallel GPUs = 1 | Total batch size (1 x 4 x 1) = 4
 "-____-"     Trainable parameters = 3,981,312 of 20,918,738,496 (0.02% trained)


Unsloth: Will smartly offload gradients to save VRAM!


**Reasoning**:
The training process has completed, as indicated by the successful execution of `trainer.train()` in the previous step. The next instruction in the subtask is to display the final memory and time statistics to understand the resource utilization and duration of the training.



In [23]:
used_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
used_memory_for_lora = round(used_memory - start_gpu_memory, 3)
used_percentage = round(used_memory / max_memory * 100, 3)
lora_percentage = round(used_memory_for_lora / max_memory * 100, 3)
print(f"{trainer_stats.metrics['train_runtime']} seconds used for training.")
print(
    f"{round(trainer_stats.metrics['train_runtime']/60, 2)} minutes used for training."
)
print(f"Peak reserved memory = {used_memory} GB.")
print(f"Peak reserved memory for training = {used_memory_for_lora} GB.")
print(f"Peak reserved memory % of max memory = {used_percentage} %.")
print(f"Peak reserved memory for training % of max memory = {lora_percentage} %.")

876.7785 seconds used for training.
14.61 minutes used for training.
Peak reserved memory = 12.816 GB.
Peak reserved memory for training = 0.0 GB.
Peak reserved memory % of max memory = 86.941 %.
Peak reserved memory for training % of max memory = 0.0 %.


## Run Inference

### Subtask:
Perform inference with your finetuned model to test its performance on new inputs.


### Instructions
1. Define a list of messages, including a system message with `reasoning language: French` and a user query like `Solve x^5 + 3x^4 - 10 = 3.`.
2. Apply the chat template to these messages using `tokenizer.apply_chat_template`, ensuring `add_generation_prompt = True`, `return_tensors = "pt"`, `return_dict = True`, and set `reasoning_effort = "medium"`. Move the inputs to the GPU (`.to("cuda")`).
3. Generate text using the `model.generate` method, passing the prepared `inputs`, setting `max_new_tokens = 64`, and using `streamer = TextStreamer(tokenizer)` to display the output as it's generated.

**Reasoning**:
The subtask is to perform inference. The instructions have been provided. The next logical step is to execute the code that performs the inference, which is available in the original notebook's cell `RdVCmTuBnVgl`.



In [24]:
messages = [
    {"role": "system", "content": "reasoning language: French\n\nYou are a helpful assistant that can solve mathematical problems."},
    {"role": "user", "content": "Solve x^5 + 3x^4 - 10 = 3."},
]
inputs = tokenizer.apply_chat_template(
    messages,
    add_generation_prompt = True,
    return_tensors = "pt",
    return_dict = True,
    reasoning_effort = "medium",
).to("cuda")
from transformers import TextStreamer
_ = model.generate(**inputs, max_new_tokens = 64, streamer = TextStreamer(tokenizer))

<|start|>system<|message|>You are ChatGPT, a large language model trained by OpenAI.
Knowledge cutoff: 2024-06
Current date: 2025-12-28

Reasoning: medium

# Valid channels: analysis, commentary, final. Channel must be included for every message.
Calls to these tools must go to the commentary channel: 'functions'.<|end|><|start|>developer<|message|># Instructions

reasoning language: French

You are a helpful assistant that can solve mathematical problems.<|end|><|start|>user<|message|>Solve x^5 + 3x^4 - 10 = 3.<|end|><|start|>assistant<|channel|>analysis<|message|>The user says "Solve x^5 + 3x^4 - 10 = 3." So we need to solve for x. It's an equation: x^5 + 3x^4 - 10 = 3. We rearrange: x^5 + 3x^


## Save Finetuned Model

### Subtask:
Save the finetuned LoRA adapters locally within the Colab environment using `model.save_pretrained`. You can later download this model or push it to Hugging Face Hub from Colab.


**Reasoning**:
The subtask requires saving the finetuned LoRA adapters locally. The instructions specify using `model.save_pretrained()` to a directory named `finetuned_model`, which directly corresponds to the existing code in the notebook.



In [25]:
model.save_pretrained("finetuned_model")
# model.push_to_hub("hf_username/finetuned_model", token = "hf_...") # Save to HF

**Reasoning**:
The subtask requires saving the finetuned LoRA adapters locally. The instructions specify using `model.save_pretrained()` to a directory named `finetuned_model`, which directly corresponds to the existing code in the notebook.



In [26]:
model.save_pretrained("finetuned_model")
# model.push_to_hub("hf_username/finetuned_model", token = "hf_...") # Save to HF

**Reasoning**:
The subtask requires saving the finetuned LoRA adapters locally. The instructions specify using `model.save_pretrained()` to a directory named `finetuned_model`, which directly corresponds to the existing code in the notebook.



In [27]:
model.save_pretrained("finetuned_model")
# model.push_to_hub("hf_username/finetuned_model", token = "hf_...") # Save to HF

## Final Task

### Subtask:
Summarize the end-to-end process of finetuning an open-source model with Unsloth and your custom data on Google Colab, highlighting the ease of use and computational benefits.


## Summary:

### Data Analysis Key Findings

*   **Model Loading Challenges and Resolution**: Initial attempts to load the `unsloth/gpt-oss-20b` base model failed due to `ValueError` (insufficient GPU RAM). This was successfully resolved by configuring `BitsAndBytesConfig` with `llm_int8_enable_fp32_cpu_offload=True` to enable 32-bit CPU offloading for modules and passing it to `FastLanguageModel.from_pretrained` along with `device_map="auto"`.
*   **LoRA Adapter Setup**: LoRA adapters were successfully applied to the model using `FastLanguageModel.get_peft_model`, configuring it for parameter-efficient fine-tuning with settings like `r=8` and `use_gradient_checkpointing="unsloth"`.
*   **Custom Data Preparation**: A custom dataset (`nz_sovereign_train_v1.jsonl`) was loaded and preprocessed using a `formatting_prompts_func` and the tokenizer, resulting in a dataset of 184 rows with a single 'text' feature, ready for training.
*   **Training Execution and Debugging**: The model was successfully trained for 30 steps using `trl.SFTTrainer`. During training setup, a `ZeroDivisionError` related to chat template matching was resolved by correcting the `response_part` in `gpt_oss_kwargs`. A persistent `AttributeError` during `train_on_responses_only` application was resolved by manually tokenizing and padding the dataset *before* passing it to the utility.
*   **Resource Utilization During Training**: The training process on a Tesla T4 GPU (max 14.741 GB memory) utilized 14.61 minutes and reached a peak reserved memory of 12.816 GB, representing 86.941% of the GPU's maximum memory.
*   **Inference Performance Discrepancy**: While inference executed successfully, the finetuned model failed to adhere to the `reasoning language: French` instruction provided in the system message during inference, generating output in English instead.
*   **Model Saving**: The finetuned LoRA adapters were successfully saved locally to a directory named "finetuned\_model".

### Insights or Next Steps

*   The model's failure to follow language instructions in the system prompt during inference indicates a potential need for further fine-tuning on diverse instructional prompts or an evaluation of the chat template's effectiveness in conveying such instructions.
*   Explore more advanced training techniques or dataset augmentations to improve the model's adherence to nuanced instructions, especially regarding output language or specific reasoning formats.
